enhancements/monitoring/alert-overrides.md

```diff
@@ -0,0 +1,280 @@
+  namespace: openshift-monitoring
+data:
+  overrides.yaml: |
+    overrides: # List of platform rule overrides.
+    - key:
+        name: "Watchdog" # Alerts are identified by name and severity tuple.
+        severity: "none"
+      action: "patch"
+      labels:
+        patched: true
+    - key:
+        name: "AlertmanagerReceiversNotConfigured"
+        severity: "Warning"
+      action: "drop"
+
+    rules: # List of additional alerting rules.
+    - alert: "CustomAlert"
+      expr: "vector(1)"
+```
+
+The controller maintains an index of alerting rule name-and-severity pairs to
+`PrometheusRule` objects, ignoring any in non-platform namespaces, i.e. any
+namespace without the `openshift.io/cluster-monitoring=true` label.  When the
+configuration changes, the `PrometheusRule` containing the targeted alert is
+found, and the alert extracted.
+
+For each override a new alerting rule is generated with the overrides applied on
+top of the existing rule, and this new rule is added to the "alert-overrides"
+`PrometheusRule` object.  Additional user-supplied rules are also added here.
+All generated alerts will be given an identifying label -- currently:
+`monitoring_openshift_io__alert_override=true`
+
+This leaves the original alerting rules present and untouched.  These, as well
+as any rules listed with the `drop` action, must be ignored by the platform
+Prometheus instance.  A `Secret` containing [alert_relabel_configs][1] data for
+Prometheus is generated to accomplish this:
+
+```yaml
+# Alerts to be dropped are matched against name, severity, and an empty
+# override label.  The empty override label differentiates the original
+# and generated alerting rules.
+
+- source_labels: [alertname, severity, monitoring_openshift_io__alert_override]
+  regex: “Watchdog;none;”
+  action: drop
+- source_labels: [alertname, severity, monitoring_openshift_io__alert_override]
+  regex: “AlertmanagerReceiversNotConfigured;Warning;”
+  action: drop
+```
+
+Note that this prevents these alerts, whether overridden or dropped, from being
+sent to Alertmanager, but they will still show up as active alerts in the
+Prometheus API.  The OpenShift console will have to be made aware of the
+labeling in order to hide these by default.
+
+### User Stories
+
+- As an OpenShift user, I want to raise the severity of an alert from `Warning`
+  to `Critical`.
+- As an OpenShift user, I want to permanently disable an alert that I know does
+  not pertain to my deployment.
+- As an OpenShift user, I want to add labels or annotations to alerts to aide in
+  routing and triage.
+- As an OpenShift user, I want to rewrite the PromQL expression or duration for
+  an alert in order to adjust thresholds to suit my environment.
+- As an OpenShift user, I want a supported way to add custom alerts using
+  metrics from components in the `openshift-*` namespaces.
+
+### API Extensions
+
+None, but this is an open question.  Is there good reason to make this a CRD
```

Comment:
+1 to CRD, could build status into suppress/changed alerts so as a user I have a summary of what I have changed?

Reply:
A CRD also provides better UX since it has built-in validation and documentation mechanisms

Reply:
Agreed. I've added a proposed structure for a CRD now.